rustls/src/conn.rs

```diff
@@ -1080,17 +1086,29 @@ impl ConnectionCommon {
     }
 }
 
-pub(crate) trait HandleState: Sized {
-    type Data;
-    type Config;
-
+pub(crate) trait State<Data, Config>: Send + Sync {
```

Comment:
It doesn't make sense for this to be in conn.rs, does it? It isn't about connection state, but about handshake state.

I suggest we move this into a new `hs` top-level module.

Further, I suggest we start following the normal Rust convention where modules are used more frequently to control visibility, which results in types and traits usually being defined in many small submodules. In this case, I think this trait should be `hs::state::State` in src/hs/state.rs.

Also, it seems like this should be:
```
trait State<C> : Send + Sync {
  fn handle(
        self: Box<Self>,
        cx: &mut C,
        message: Message,
    ) -> Result<Box<dyn State<C>>, Error>;
  }

  fn export_keying_material(...) // same as you have it.
}
```

Then the implementations would be `impl hs::State<ClientContext>` and `impl hs::State<ServerContext>`.

Reply:
`Traffic` is also a state in the state machine. The `State` trait is about how the 1.2/1.3 client/server code interacts with the shared logic (buffers, encryption/decryption); so, in my view, `State` is an interface for `ConnectionCommon` to abstract over all those aspects of a particular connection. That said, I don't mind moving it somewhere else, but I think might be a little bit confusing to restrict to just `hs` since it's also used for post-handshake key updates.

In the projects I've worked on, modules are not used much more frequent than in current rustls to control visibility. Personally adding a module which just contains one item seems like unnecessary complexity to me.

Reply:
OK, thanks for explaining. Seems reasonable.